### PR TITLE
Interactivity API refactor to TypeScript (utils & kebabToCamelCase)

### LINKED ISSUE #60147: Interactivity API refactor to TypeScript (utils & kebabToCamelCase)
## What problem does this address?
Actually the Interactivity API is written on JavaScript, without any type annotation. Should be refactored to TypeScript to use all features from that language.

## What is your proposed solution?
Refactor all the API files to TypeScript. This issue is for addressing the utils and kebabToCamelCase

### PR DESCRIPTION
**Pull request by garridinsi**
## What?

Fixes #60147

## Why?

To profit from the TypeScript features.

## How?

Refactors the js files to ts and adds type annotation

### CODE REVIEW

File: packages/interactivity/src/utils.ts [hunk 1]
```diff
@@ -203,21 +228,31 @@ export function useCallback( callback, inputs ) {
...
 
-// For wrapperless hydration.
-// See https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
-export const createRootFragment = ( parent, replaceNode ) => {
+/**
+ * Creates a root fragment by replacing a node or an array of nodes in a parent element.
+ * For wrapperless hydration.
+ * See https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
+ *
+ * @param parent      - The parent element where the nodes will be replaced.
```
Comment (gziolo) [inline packages/interactivity/src/utils.ts]: Here and in other places the hyphen isn’t necessary:
```suggestion
 * @param parent      The parent element where the nodes will be replaced.
```

File: packages/interactivity/src/utils.ts [hunk 2]
```diff
@@ -203,21 +228,31 @@ export function useCallback( callback, inputs ) {
...
+ * @return The created root fragment.
+ */
+export const createRootFragment = (
+	parent: ParentElement,
+	replaceNode: Node | Node[]
+) => {
 	replaceNode = [].concat( replaceNode );
 	const s = replaceNode[ replaceNode.length - 1 ].nextSibling;
-	function insert( c, r ) {
+	function insert( c: any, r: any ) {
```
Comment (gziolo) [inline packages/interactivity/src/utils.ts]: Optional task: it would be helpful to unfold what `c`, `r` and `s` mean.

File: packages/interactivity/src/utils.ts [hunk 3]
```diff
@@ -51,20 +66,30 @@ function createFlusher( compute, notify ) {
...
+ * infinite loops: https://github.com/preactjs/signals/pull/290
+ *
+ * @param callback - The callback function to be executed.
+ */
+export function useSignalEffect( callback: Function ) {
 	_useEffect( () => {
 		let eff = null;
 		let isExecuting = false;
+
+		/**
```
Comment (gziolo) [inline packages/interactivity/src/utils.ts]: One line comments can be shortened to:

```js
// Notifies the callback function to execute after the next frame.
```

**Review by gziolo — COMMENTED** [review]
Nice improvements adding missing types and JSDoc comments.

**Comment by gziolo** [comment]
I see now that @c4rl0sbr4v0 has an alternative proposal open for adding TypeScript types to the same package but at a larger scale: https://github.com/WordPress/gutenberg/pull/58718. So it wil require some coordination 😀

**Comment by garridinsi** [comment]
> I see now that @c4rl0sbr4v0 has an alternative proposal open for adding TypeScript types to the same package but at a larger scale: #58718. So it wil require some coordination 😀

Hi! Yes! @c4rl0sbr4v0 made another PR but now is outdated as of some Interactivity API changes. This PR was made on WordCamp Torrelodones 2024 Contributor Day, and @c4rl0sbr4v0 is aware of this, as he was the Core table leader

**Comment by sirreal** [comment]
Thanks for this work! I would like to review this but need a few days to get back to it. I see a few places we can improve and/or simplify things.

**Comment by garridinsi** [comment]
Thanks @sirreal! 

Sorry @gziolo for the Review request, was a miss click on my end

**Comment by cbravobernal** [comment]
> I see now that @c4rl0sbr4v0 has an alternative proposal open for adding TypeScript types to the same package but at a larger scale: #58718. So it wil require some coordination 😀

Don't worry, I can handle that coordination :-)

File: packages/interactivity/src/utils.ts (hunk 2 above)
Comment (cbravobernal) [reply packages/interactivity/src/utils.ts]: As far as I remember,
 - c is children.
 - r is root.
 - s is sibling.
 
 cc @DAreRodz, correct me if I'm wrong.

File: packages/interactivity/src/utils.ts [hunk 4]
```diff
@@ -20,9 +20,16 @@ import {
...
 
-const afterNextFrame = ( callback ) => {
-	return new Promise( ( resolve ) => {
+/**
+ * Executes a callback function after the next frame is rendered.
+ *
+ * @param callback - The callback function to be executed.
+ * @return A promise that resolves after the callback function is executed.
+ */
+const afterNextFrame = ( callback: Function ) => {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: `Function` isn't an ideal type for this, [it's better to express the callback type](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#callback-types):

```suggestion
const afterNextFrame = ( callback: () => void ) => {
```

File: packages/interactivity/src/utils.ts [hunk 5]
```diff
@@ -36,12 +43,20 @@ const afterNextFrame = ( callback ) => {
...
+ * Using the mangled properties:
+ * this.c: this._callback
+ * this.x: this._compute
+ * https://github.com/preactjs/signals/blob/main/mangle.json
+ *
+ * @param compute The function that computes the value to be flushed.
+ * @param notify  The function that notifies listeners when the value is flushed.
+ * @return The Flusher object with `flush` and `dispose` properties.
+ */
+function createFlusher( compute: Function, notify: () => void ): Flusher {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: I think this compute type is correct and we can let the object be inferred:

```suggestion
function createFlusher( compute: () => unknown, notify: () => void ) {
```

File: packages/interactivity/src/utils.ts [hunk 6]
```diff
@@ -51,20 +66,30 @@ function createFlusher( compute, notify ) {
 	return { flush, dispose };
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: This is a nice spot for a const assertion.
```suggestion
	return { flush, dispose } as const;
```

File: packages/interactivity/src/utils.ts [hunk 7]
```diff
@@ -51,20 +66,30 @@ function createFlusher( compute, notify ) {
...
+ * @param callback The callback function to be executed.
+ */
+export function useSignalEffect( callback: Function ) {
 	_useEffect( () => {
 		let eff = null;
 		let isExecuting = false;
+
+		/**
+		 * Notifies the callback function to execute after the next frame.
+		 */
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: I'd remove this comment entirely.
```suggestion
```

File: packages/interactivity/src/utils.ts [hunk 8]
```diff
@@ -51,20 +66,30 @@ function createFlusher( compute, notify ) {
...
-export function useSignalEffect( callback ) {
+/**
+ * Custom hook that executes a callback function whenever a signal is triggered.
+ * Version of `useSignalEffect` with a `useEffect`-like execution. This hook
+ * implementation comes from this PR, but we added short-cirtuiting to avoid
+ * infinite loops: https://github.com/preactjs/signals/pull/290
+ *
+ * @param callback The callback function to be executed.
+ */
+export function useSignalEffect( callback: Function ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
export function useSignalEffect( callback: () => unknown ) {
```

File: packages/interactivity/src/utils.ts [hunk 9]
```diff
@@ -75,17 +100,17 @@ export function useSignalEffect( callback ) {
  * accessible whenever the function runs. This is primarily to make the scope
  * available inside hook callbacks.
  *
- * @param {Function} func The passed function.
- * @return {Function} The wrapped function.
+ * @param func The passed function.
+ * @return The wrapped function.
  */
-export const withScope = ( func ) => {
+export const withScope = ( func: Function ): ( () => void ) => {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: This function is difficult to type well. The trick is that it transforms a generator function to an async function that takes the generator args and returns the final iterator result value.

I'm not sure what's best for now, maybe just keep it wide open with `func` and return types `(...args: any[]) => any`.

File: packages/interactivity/src/utils.ts [hunk 10]
```diff
@@ -125,9 +150,9 @@ export const withScope = ( func ) => {
  * This hook makes the element's scope available so functions like
  * `getElement()` and `getContext()` can be used inside the passed callback.
  *
- * @param {Function} callback The hook callback.
+ * @param callback The hook callback.
  */
-export function useWatch( callback ) {
+export function useWatch( callback: Function ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
export function useWatch( callback: () => void ) {
```

File: packages/interactivity/src/utils.ts [hunk 11]
```diff
@@ -138,9 +163,9 @@ export function useWatch( callback ) {
  * This hook makes the element's scope available so functions like
  * `getElement()` and `getContext()` can be used inside the passed callback.
  *
- * @param {Function} callback The hook callback.
+ * @param callback The hook callback.
  */
-export function useInit( callback ) {
+export function useInit( callback: Function ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: If we an import for `type EffectCallback` from `preact`, we can use that where `_useEffect` is called:

```suggestion
export function useInit( callback: EffectCallback ) {
```

File: packages/interactivity/src/utils.ts [hunk 12]
```diff
@@ -152,12 +177,12 @@ export function useInit( callback ) {
...
- *                            function.
- * @param {any[]}    inputs   If present, effect will only activate if the
- *                            values in the list change (using `===`).
+ * @param callback Imperative function that can return a cleanup
+ *                 function.
+ * @param inputs   If present, effect will only activate if the
+ *                 values in the list change (using `===`).
  */
-export function useEffect( callback, inputs ) {
+export function useEffect( callback: Function, inputs: any[] ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
export function useEffect( callback: EffectCallback, inputs: any[] ) {
```

File: packages/interactivity/src/utils.ts [hunk 13]
```diff
@@ -169,12 +194,12 @@ export function useEffect( callback, inputs ) {
...
- *                            function.
- * @param {any[]}    inputs   If present, effect will only activate if the
- *                            values in the list change (using `===`).
+ * @param callback Imperative function that can return a cleanup
+ *                 function.
+ * @param inputs   If present, effect will only activate if the
+ *                 values in the list change (using `===`).
  */
-export function useLayoutEffect( callback, inputs ) {
+export function useLayoutEffect( callback: Function, inputs: any[] ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
export function useLayoutEffect( callback: EffectCallback, inputs: any[] ) {
```

File: packages/interactivity/src/utils.ts [hunk 14]
```diff
@@ -203,21 +228,31 @@ export function useCallback( callback, inputs ) {
...
- * @param {any[]}    inputs  If present, effect will only activate if the
- *                           values in the list change (using `===`).
+ * @param factory Imperative function that can return a cleanup
+ *                function.
+ * @param inputs  If present, effect will only activate if the
+ *                values in the list change (using `===`).
  */
-export function useMemo( factory, inputs ) {
+export function useMemo( factory: Function, inputs: any[] ) {
 	_useMemo( withScope( factory ), inputs );
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: I think this function is completely wrong, fixing it here: https://github.com/WordPress/gutenberg/pull/60474

Let's revert these changes for now.

File: packages/interactivity/src/utils.ts [hunk 15]
```diff
@@ -186,12 +211,12 @@ export function useLayoutEffect( callback, inputs ) {
...
- * @param {any[]}    inputs   If present, effect will only activate if the
- *                            values in the list change (using `===`).
+ * @param callback Imperative function that can return a cleanup
+ *                 function.
+ * @param inputs   If present, effect will only activate if the
+ *                 values in the list change (using `===`).
  */
-export function useCallback( callback, inputs ) {
+export function useCallback( callback: Function, inputs: any[] ) {
 	_useCallback( withScope( callback ), inputs );
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: I think this function is completely wrong, fixing it here: https://github.com/WordPress/gutenberg/pull/60474

Let's revert these changes for now.

File: packages/interactivity/src/types.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+export interface Flusher {
+	flush: () => void;
+	dispose: () => void;
+}
+export interface ParentElement extends Element {
```
Comment (sirreal) [inline packages/interactivity/src/types.ts]: I'd delete this entire file and define these types elsewhere. They're really internal types and can live in the same module.

File: packages/interactivity/src/utils.ts [hunk 17]
```diff
@@ -203,21 +228,31 @@ export function useCallback( callback, inputs ) {
...
+ * Creates a root fragment by replacing a node or an array of nodes in a parent element.
+ * For wrapperless hydration.
+ * See https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
+ *
+ * @param parent      The parent element where the nodes will be replaced.
+ * @param replaceNode The node or array of nodes to replace in the parent element.
+ * @return The created root fragment.
+ */
+export const createRootFragment = (
+	parent: ParentElement,
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: We're doing some weird stuff below, and we don't want to be passed that weird type but just want an `Element`.

```suggestion
	parent: Element,
```

Below, where we know we're doing some weird stuff, I think we're justified in using an `any`.

File: packages/interactivity/src/utils.ts [hunk 18]
```diff
@@ -203,21 +228,31 @@ export function useCallback( callback, inputs ) {
...
+	replaceNode: Node | Node[]
+) => {
 	replaceNode = [].concat( replaceNode );
 	const s = replaceNode[ replaceNode.length - 1 ].nextSibling;
-	function insert( c, r ) {
+	function insert( c: any, r: any ) {
 		parent.insertBefore( c, r || s );
 	}
 	return ( parent.__k = {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
	return ( ( parent as any ).__k = {
```

File: packages/interactivity/src/utils/kebab-to-camelcase.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Transforms a kebab-case string to camelCase.
+ *
+ * @param str The kebab-case string to transform to camelCase.
+ * @return     The transformed camelCase string.
```
Comment (sirreal) [inline packages/interactivity/src/utils/kebab-to-camelcase.ts]:
```suggestion
 * @return The transformed camelCase string.
```

File: packages/interactivity/src/utils.ts [hunk 20]
```diff
@@ -20,9 +20,16 @@ import {
 	setNamespace,
 	resetNamespace,
 } from './hooks';
+import type { Flusher, ParentElement } from './types';
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]:
```suggestion
```

**Review by sirreal — COMMENTED** [review]
Thanks for the patience @garridinsi, I've left a number of suggestions and pushed a small change. I'm happy to push some of these myself or discuss further.

File: packages/interactivity/src/utils.ts (hunk 2 above)
Comment (sirreal) [reply packages/interactivity/src/utils.ts]: It's all from https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c, I think @c4rl0sbr4v0 is probably right.

Comment (cbravobernal) [reply packages/interactivity/src/utils.ts]: There is a TypeScripted version in case you need:

https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c?permalink_comment_id=4558744#gistcomment-4558744

File: packages/interactivity/src/types.ts (hunk 16 above)
Comment (cbravobernal) [reply packages/interactivity/src/types.ts]: I think it is on that file following the repository structure. But we can create them only for external types.

**Comment by cbravobernal** [comment]
@sirreal I updated with all the changes you suggested. Feel free to check it again. Thanks!

File: packages/interactivity/src/utils.ts (hunk 9 above)
Comment (sirreal) [reply packages/interactivity/src/utils.ts]: I ended up pushing a proper typing for this.

File: packages/interactivity/src/utils.ts [hunk 21]
```diff
@@ -75,17 +103,30 @@ export function useSignalEffect( callback ) {
...
+	Func extends ( ...args: any[] ) => Generator< any, any >,
+>(
+	func: Func
+): (
+	...args: Parameters< Func >
+) => ReturnType< Func > extends Generator< any, infer Return >
+	? Promise< Return >
+	: never;
+export function withScope< Func extends Function >( func: Func ): Func;
+export function withScope( func ) {
```
Comment (sirreal) [inline packages/interactivity/src/utils.ts]: I changed this to be a `function` so I could use function overloading for the typing here.

[See types demo in practice](https://www.typescriptlang.org/play?target=99&jsx=0#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwO6YwAWAymhGMADwBQc9cAYqmnKDMCgCYDOcAFADphAQygBzHgC44IlAE8A2gF0AlHAC8APjgBxTsCgiY0KnPkAaWQq0WaW-snQzm6VTKGiJ0uAAUxIgC2wBxQPFSuaFrq2nAASiFIUCgAKvKUESw67Jy8egZGJlBmClaYKAiG8YnJOgD8flAQgZg81AkwSSg6MijAAG6GANw0oJCwiCxYOPiEpOQZkWwgHNx8kdPdjiwuLO5MLCNj0PBOGNi4BMRkFMDbzgLCgmKSMqgA1igQeCgqMTofL4-dQAbxodAYmAQAjOdUE5BQPBgUCQGGgcJQQWAmg0GjgAHJ9H1CtANhd8aCIQwGFAargRDx5OhHl5Xn4AsFQuEYOlgBBoWdopodGCqdTxQikXBxJxNJN0J5nt51Az8sTjNARuLtXAADYhOD9ES6pDAGTmLU66n6+CEc0KS1W+h4IiYfUCZGmylip0MZHyOBg32+whymUoQR9Fb8I0m4CqR3B+gAX0Q5WNuoDoqT1OTPuD-sD+ZzhuNprlIjwIlDhEEsdNiaTqbQxjQRAE8aLJe14cExCaeH48cbwbz3bgUIEta4OE7ACNaSJ3iPtWPg7TOslS3GV3Bk4m1-QN10WUq2f4jJzDNzefz5VF-vfFS8eAmaHmaJLTnKrvNbvdzhwAAqARQTgeRMGAXUuEaZpWmAQRaR4CBdUGfgAEYEzgY8t09bFkzfAB6QjqQAPTqT8cClBAACYfzmG5KAAzZQMDbC6TgPC9yIkiGHImggA)

**Review by sirreal — APPROVED** [review]
I think this is ready, thanks folks!
